File: LongestPalindromicSubstring.java
/*
Runtime: 25 ms, faster than 94.76% of Java online submissions for Longest Palindromic Substring.
Memory Usage: 41.8 MB, less than 99.07% of Java online submissions for Longest Palindromic Substring.
*/
// Optimized Solution: O(n^2)
class Solution {
    public String longestPalindrome(String s) {

        int start = 0;
        int end = 0;
        for(int i=0; i<s.length(); i++){
            int len1 = findPalindrome(s,i,i);
            int len2 = findPalindrome(s,i,i+1);
            int maxLength = Math.max(len1, len2);
            if(maxLength > end - start){
                start = i - (maxLength-1)/2;
                end = i + maxLength/2;
            }
        }
        return s.substring(start, end+1);
    }
    
    public static int findPalindrome(String s, int left, int right){
        
        while(left>=0 && right<s.length() && s.charAt(left)==s.charAt(right)){
            left--;
            right++;
        }
        return right - left - 1;
    }
    
}

/*
//Brute Force Solution - O(n^3)
class Solution {
    public String longestPalindrome(String s) {
        int last = s.length()-1;
        String longestPalindrome = "";
        for(int i=0; i<s.length(); i++){
            for(int j = s.length()-1; j>=i; j--){
                StringBuilder str = new StringBuilder();
                String temp = s.substring(i,j+1);
                str.append(temp);
                
                str = str.reverse();
                if(temp.equals(str.toString())){
                    if(longestPalindrome.length() < temp.length()){
                        longestPalindrome = temp;
                    }
                }
            }
        }
        return longestPalindrome;
    }
}
*/

File: LongestPalindromicSubstring.js
/**
 * @param {string} s
 * @return {string}
 */
// TC: O(N^2) SC: O(1)
const palindromeLength = (s, left, right) => {
    while(left >= 0 && right < s.length && s[left] === s[right]){
        left--;
        right++;
    }
    return right - left - 1;
}

var longestPalindrome = function(s) {
    let start = 0, end = 0;
    for(let i=0; i<s.length; i++){
        const len1 = palindromeLength(s, i, i);
        const len2 = palindromeLength(s, i, i+1);
        const len = Math.max(len1, len2);
        if(end - start < len){
            start = i - Math.floor((len - 1)/2);
            end = i + Math.floor(len/2);
        }
    }
    return s.slice(start, end + 1);
};


// Dynamic Programming TC: O(N^2) SC: O(N^2)
var longestPalindrome = function(s) {
    // Initialization
    const dp = []; // Create the 2D table 'dp' to store palindrome information
    for (let i = 0; i < s.length; i++) {
        dp[i] = []; 
        for (let j = 0; j < s.length; j++) {
            dp[i][j] = false; // Initially mark all entries in the table as 'false' (not palindromes)
        }
    }

    let start = 0, end = 0; // Variables to track the starting and ending indices of the longest palindrome

    // Filling the 'dp' table in a bottom-up fashion
    for (let i = s.length - 1; i >= 0; i--) { // Outer loop iterates in reverse for easier processing
        for (let j = i; j < s.length; j++) { 
            
            // Two conditions for a substring to be a palindrome:
            if (s[i] === s[j] && j - i <= 2) {  // Condition 1: Direct match, and substring is of length 3 or less
                dp[i][j] = true;
            } else if (s[i] === s[j] && dp[i + 1][j - 1]) { // Condition 2: Direct match and the inner substring is already a palindrome
                dp[i][j] = true; 
            } else {
                dp[i][j] = false; // If the conditions don't hold, not a palindrome
            }

            // Update the longest palindrome if necessary
            if (dp[i][j] && (j - i + 1 > end - start + 1)) {
                start = i;
                end = j;
            }
        }
    }

    // Extract and return the result
    return s.slice(start, end + 1); 
};
